test/util/http.js

```diff
@@ -569,61 +570,68 @@ class MockHttp {
-          this._checkStateBeforeRequest();
-          return this._request(this._component);
-        })
-        // Wait for any responses to be processed.
-        .finally(wait)
-        .finally(pollWork != null
-          ? () => waitUntil(() => pollWork(this._component))
-          : noop)
-        .finally(() => { this._container.http.respond(null); })
-        .then(() => { this._checkStateAfterWait(); })
-        .then(() => callback(this._component))
-        .then(result => ({ component: this._component, result })));
-    return new MockHttp({ container: this._container, previousPromise: promise });
+    if (this._location != null && this._container.router == null)
+      throw new Error('container does not contain a router');
+    return new MockHttp({
+      container: this._container,
+      previousPromise: this._executeSeries(optionsOrCallback)
+    });
   }
 
   afterResponse(optionsOrCallback) {
     return this.afterResponses(optionsOrCallback);
   }
 
-  complete() { return this.afterResponses(component => component); }
-
-  // Sets up `http` in the container and sets properties of the MockHttp that
-  // are used within this.afterResponses(). this._initialPromise() does nothing
-  // if this._previousPromise is rejected.
-  _initialPromise() {
-    const promise = this._previousPromise != null
-      ? this._previousPromise
-      : Promise.resolve({});
-    // `component` is the component that the previous promise mounted (if any).
-    return promise.then(({ component }) => {
-      this._container.http.respond(this._http());
+  complete() { return this.afterResponses(identity); }
+
+  async _executeSeries({ callback, pollWork = undefined }) {
+    if (this._previousPromise != null) {
+      // Immediately reject if this._previousPromise is rejected. `component` is
+      // the component that the previous promise mounted (if any).
+      const { component } = await this._previousPromise;
       this._component = component;
-      this._requestCount = 0;
-      this._errorFromBeforeAnyResponse = null;
-      this._errorFromBeforeEachResponse = null;
-      this._errorFromResponse = null;
-      this._requestResponseLog = [];
-    });
+    }
+
+    this._requestCount = 0;
+    this._errorFromBeforeAnyResponse = null;
+    this._errorFromBeforeEachResponse = null;
+    this._errorFromResponse = null;
+    this._requestResponseLog = [];
+    const { http, router } = this._container;
+    http.respond(this._http());
+
+    this._errorFromRouter = null;
+    const removeHandler = router != null
+      ? router.onError(error => { this._errorFromRouter = error; })
+      : noop;
+
+    try {
+      if (this._location != null) await router.push(this._location);
+      if (this._mount != null) this._component = this._mount();
+      const mountTriggeredNavigation = this._mount != null && router != null &&
+        router.currentRoute.value === START_LOCATION;
+      if (mountTriggeredNavigation) await router.isReady();
```

Comment:
Instead:

```js
if (router != null) await router.isReady();
```

\+ add a comment about how `this._mount()` can trigger navigation.